src/core/session.ts

```diff
@@ -296,8 +296,8 @@ export class Session
 
   // Application events
 
-  applicationAllowsFollowingLinkToLocation(link: Element, location: URL) {
-    const event = this.notifyApplicationAfterClickingLinkToLocation(link, location)
+  applicationAllowsFollowingLinkToLocation(link: Element, location: URL, ev: MouseEvent) {
+    const event = this.notifyApplicationAfterClickingLinkToLocation(link, location, ev)
     return !event.defaultPrevented
```

Comment:
As an alternative to dispatching the event with a reference to the original, would explicitly preventing the default if `event.defaultPrevented` serve the same purpose without expanding the surface area of the `CustomEvent.detail`?

```typescript
applicationAllowsFollowingLinkToLocation(link: Element, location: URL, mouseEvent: MouseEvent) {
  const event = this.notifyApplicationAfterClickingLinkToLocation(link, location)

  if (event.defaultPrevented) {
    mouseEvent.preventDefault()
    return false
  } else {
    return true
  }
}
```

Reply:
the use cases would be:

1. I want to prevent Turbo from navigating and let the browser do it -> prevent `turbo:click`
2. I want to prevent all navigations -> prevent both events

Reply:
With that in mind, could we make Frame navigation internally aware of the cancellation?

Reply:
from my understanding, this is where we trigger the Frame navigation:

https://github.com/hotwired/turbo/blob/98cdc4037788f078a741c7fb50412db0101bcb97/src/core/frames/link_interceptor.ts#L37-L43

this is listening to the `turbo:click` event, so it _should_ be possible for it to know if the event is `defaultPrevented`. The problem is that the interceptor runs **before** any `addEventListener` I add, so the it always evaluates the event first, and it is never prevented.

maybe we could update 

https://github.com/hotwired/turbo/blob/98cdc4037788f078a741c7fb50412db0101bcb97/src/core/session.ts#L153-L159

to dispatch another event (if the click isn't prevented) that the LinkInterceptor listens to

Reply:
@seanpdoyle would adding another _internal_ event make sense here?